### ISSUE TITLE
Correctly handle menu_calling (ESC Menu) and encounter_calling flags (Random Encounter)

### ISSUE THREAD

**Issue by fmatthew5876** [EasyRPG/Player#1691]
We are not frame accurate with our handling of `menu_calling` and `encounter_calling`

`menu_calling`

- [x] Trigger ESC key to `menu_calling` at the right time within frame and under the right conditions
- [x] What gets blocked when `menu_calling` is active
- [x] What clears `menu_calling`
- [x] Frame accurate summoning of the menu

`encounter_calling`

- [x] Trigger `encounter_calling` on the right frame of movement with random encounters
- [x] Trigger `encounter_calling` at the right time within the frame and under the right conditions
- [x] What gets blocked when `encounter_calling` is active
- [x] What clears `encounter_calling`
- [x] Frame accurate summoning of the battle.

From Cherry:

Conditions that prevent party movement on map

    Airship is currently lifting off or landing
    Save.events.commands.length > 0 (Note that this condition will, if fulfilled, clear menu_calling before aborting)
    Message window is visible
    Party has move_route_overwrite
    Party has pause (irrelevant)
    encounter_calling flag was set and was processed now
    menu_calling flag was set and was processed now
    Any common or map event has waiting_execution (so you are right about that!)

Conditions that prevent decision key on map

    All of the above
    Party's remaining_step > 0 (checked after potentially initiating movement from directional keys)

Conditions that prevent escape key on map

(This happens after all of the above, without any of the above conditions though)

    Menu is not allowed (of course)
    Message window is visible
    Save.events.commands.length > 0

(Note that this won't immediately invoke the menu, but instead set the menu_calling flag.)

>    Why were these flags invented instead of just changing the scene directly like the event commands do?
>
> I guess in order to give those things a lower priority than others, especially event-controlled ones, and maybe to allow events to react on ESC press too, before the menu is actually called...? I'm not sure.
>
>    When does menu_calling get checked and the menu scene called? At the end of the same frame we detected ESC key?
>
>At the next frame, at the point where my post said "menu_calling flag was set and was processed now" above.
>
>    Are there any special cases with menu_calling outside of that? Anything which will later clear the flag? How does it interact if a scene call from an event happened on the same frame?
>
>Yes, only for random encounters. It is cleared when the player position changes by something other than movement (usually teleport), when a foreground event is active, when the encount_calling was set and is being processed.
>
>    It looks like encounter_calling is only used for random encounters. Is there some frame delay on when encounter_calling is set and when the scene is changed to battle? Special caveats with this flag?
>
>It is processed at the point I wrote "encounter_calling flag was set and was processed now" in the post above. (So it has the same delay as menu_calling.) It is cleared when a foreground event is triggered.
```

**Comment by fmatthew5876** [comment]
@CherryDT

As part of the player's update routine we also animate the airship `ascent` and `descent` flag and also handle finishing ship `boarding` and `unboarding`.

Do these vehicle actions happen before or after processing input?

If they happen before, that means it would be possible to do an input action on the same frame the vehicle actions complete. If not, you'd have to wait until the next frame.

**Comment by fmatthew5876** [comment]
Test Cases 1 - encounter calling frames
--------------------------------------

1. Create map with encounter rate 1 and one battle encounter

Setup EV01 as parallel with
```
Loop:
  Var1 = Player X
  If Var1 >= 11
    OpenSaveMenu
```

Start player on X=10, step right

| Save | remaining_step | encounter_calling |
| -- | -- | -- |
| 1 | 192 | 1 |
| 2 | 160 | 1 |
| 3 | 128 | 1 |
| 4 | 96 | 1 |
| 5 | 64 | 1 |
| 6 | 32 | 1 |
| 7 | 0 | 1 |

Battle called after save 7

Result: encounter_calling flag is set when player movement begins, and triggers battle when movement finishes.

**Comment by fmatthew5876** [comment]
Fixed by #1714

**Comment by fmatthew5876** [comment]
Test Cases 2 - Airship ascent / descent and the menu
------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
Loop:
  OpenSaveMenu
```

Board the airship, and hit cancel while ascending

| Save | remaining ascent | menu calling |
| -- | -- | -- |
| 1 | 144 | 1 |
| 18 | 8 | 1 |
| 19 | 0 | 1 |
| Menu Called | | |
| 20 | 0 | 0 |

2. Board the airship, land, and hit cancel while landing

| Save | remaining descent | menu calling |
| -- | -- | -- |
| 1 | 112 | 1 |
| 14 | 8 | 1 |
| 15 | 0 | 1 |
| Menu called | | |
| 16 | 0 | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 3 - Ship boarding / unboarding and the menu
------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
Loop:
  OpenSaveMenu
```

Board the ship, and hit rapidly cancel while boarding

| Save | remaining step | aboard | menu calling |
| -- | -- | -- | -- |
| 1 | 64 | 0 | 1 |
| 2 | 32 | 0 | 1 |
| 3 | 0 | 1 | 1 |
| Menu Called | | | |
| 4 | 0 | 0 | 0 |

2. Board the ship, unboard, and rapid hit cancel while unboarding

| Save | remaining step | aboard | menu calling |
| -- | -- | -- | -- |
| 1 | 64 | 1 | 1 |
| 2 | 32 | 1 |1 |
| 3 | 0 | 0 | 1 |
| Menu called | | | |
| 4 | 0 | 0 | 0 |

**Comment by fmatthew5876** [comment]
Test Cases 4 - Messages
------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
Message: Hello
```

Start a new game and press ESC

Result: "Hello" message, menu never called.

**Comment by fmatthew5876** [comment]
Test Cases 5 - more vehicle tests
--------------------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
Board or Unboard vehicle
````

Step on the airship and press escape

Result: Airship ascends, and menu called when ascent finishes

2. Step in front of the ship, face it, and press escape

Result: Ship boarded, and menu called when boarding finishes

**Comment by fmatthew5876** [comment]
Test Cases 6 - move routes
--------------------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel, Enter, Any direction
SetMoveRoute: Player, Up
````

Press any direction

Result: Player steps up

2. Press Cancel

Result: Player steps up and menu is called after step finishes

3. Put player on same tile as airship, press enter

Result: Player steps up and does not board

4. Put ship above the player, press enter

Result: Player movement is blocked by ship, does not board.

**Comment by fmatthew5876** [comment]
Test Cases 7 - Move routes and encounters
----------------------

Setup the map with encounter rate = 1 and at least one battle.

1. Setup EV01 parallel
```
KeyInput, Wait, Any direction
SetMoveRoute: Player, Up
````

Move

Result: Player steps up, no battle occurs

2. Change EV01 to
```
KeyInput, Wait, Any direction
Wait 0.0s
SetMoveRoute: Player, Up
````

Step in any direction

Result: Player steps in requested direction, steps up, and then battle starts

**Comment by fmatthew5876** [comment]
Test Cases 8 - Menu vs walking
----------------------

1. Setup EV01 autostart
```
SetMoveRoute: Player, Speed Down, Speed Down
EraseEvent
```

Step, and then press Enter while stepping

Result: Menu is called when player finishes moving.

**Comment by fmatthew5876** [comment]
Test Cases 9 - Transitions and menu
----------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
HideScreen
```

Result: When you press cancel, the screen fades out and the menu is not called. Pressing enter again while screen is blank summons the menu.

2. Setup EV01 parallel
```
KeyInput, Wait, Cancel
HideScreen
ShowScreen
```

Result: Press cancel,screen fades out and then fades in. Menu is not callable.

Conclusion
---------------

Hide screen is a no-op when the screen is already hidden, so the menu is allowed in the first test.

**Comment by fmatthew5876** [comment]
Test Cases 10 - Battle animations and menu
---------------------

1. Setup EV01 parallel
```
KeyInput, Wait, Cancel
ShowBattleAnimation: Player
```

Press ESC

Result: Battle animation starts, and menu is immediately called. The battle animation is on frame 2 when the transition to menu starts.

**Comment by fmatthew5876** [comment]
reopened until these are covered by unit tests

**Comment by Ghabry** [comment]
Imo it is confusing to reopen tons of old issues :/
Technically they are not issues right now and you have this checklist already.
